bee-tangle/src/tangle.rs

```diff
@@ -0,0 +1,535 @@
+        })
+    }
+
+    /// Returns the children of a vertex, if we know about them.
+    pub async fn get_children(&self, message_id: &MessageId) -> Option<Vec<MessageId>> {
+        // Effectively atomic
+        self.children_inner(message_id).await.map(|approvers| approvers.clone())
+    }
+
+    /// Returns the number of children of a vertex.
+    pub async fn num_children(&self, message_id: &MessageId) -> usize {
+        // Effectively atomic
+        self.children_inner(message_id)
+            .await
+            .map_or(0, |approvers| approvers.len())
+    }
+
+    #[cfg(test)]
+    pub async fn clear(&mut self) {
+        self.vertices.write().await.clear();
+    }
+
+    // Attempts to pull the message from the storage, returns true if successful.
+    async fn pull_message(&self, message_id: &MessageId, prevent_eviction: bool) -> bool {
+        let contains_now = if prevent_eviction {
+            self.vertices.write().await.get_mut(message_id).map_or(false, |v| {
+                if v.message().is_some() {
+                    v.prevent_eviction();
+                    true
+                } else {
+                    false
+                }
+            })
+        } else {
+            self.contains_inner(message_id).await
+        };
+
+        // If the tangle already contains the tx, do no more work
+        if contains_now {
+            // Insert cache queue entry to track eviction priority
+            self.cache_queue.lock().await.put(*message_id, ());
+
+            true
+        } else if let Ok(Some((tx, metadata))) = self.hooks.get(message_id).await {
+            // Insert cache queue entry to track eviction priority
+            self.cache_queue.lock().await.put(*message_id, ());
+
+            self.insert_inner(*message_id, tx, metadata, prevent_eviction).await;
+
+            true
+        } else {
+            false
+        }
+    }
+
+    async fn perform_eviction(&self) {
+        let max_len = self.max_len.load(Ordering::Relaxed);
+        let len = self.vertices.read().await.len();
+        if len > max_len {
+            let mut vertices = self.vertices.write().await;
+            let mut cache_queue = self.cache_queue.lock().await;
+            while vertices.len() > ((1.0 - CACHE_THRESHOLD_FACTOR) * max_len as f64) as usize {
+                let remove = cache_queue.pop_lru().map(|(id, _)| id);
+
+                if let Some(message_id) = remove {
+                    if let Some(v) = vertices.remove(&message_id) {
+                        if !v.can_evict() {
+                            // Reinsert it if we're not permitted to evict it yet (because something is using it)
+                            vertices.insert(message_id, v);
+                            cache_queue.put(message_id, ());
+                        }
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+// #[cfg(test)]
```

Comment:
Should this be here?

Reply:
Going to be replacing these with a more comprehensive set of tests anyway, so I'll remove it.